contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java

```diff
@@ -0,0 +1,462 @@
+import icon.proto.core.client.Height;
+import java.math.BigInteger;
+import score.Address;
+import score.Context;
+import score.RevertedException;
+import score.UserRevertException;
+import score.UserRevertedException;
+import score.annotation.EventLog;
+import score.annotation.External;
+import score.annotation.Optional;
+import score.annotation.Payable;
+
+
+public class CallServiceImpl extends AbstractCallService {
+
+    public CallServiceImpl(Address _ibc, BigInteger _timeoutHeight) {
+        this.ibcHandler.set(_ibc);
+        admin.set(Context.getOwner());
+        this.timeoutHeight.set(_timeoutHeight);
+    }
+
+    @External
+    public void setTimeoutHeight(BigInteger _timeoutHeight) {
+        onlyAdmin();
+        this.timeoutHeight.set(_timeoutHeight);
+    }
+
+    @External(readonly = true)
+    public BigInteger getTimeoutHeight() {
+        return this.timeoutHeight.getOrDefault(BigInteger.ZERO);
+    }
+
+    private void checkCallerOrThrow(Address caller, String errMsg) {
+        Context.require(Context.getCaller().equals(caller), errMsg);
+    }
+
+    private void onlyOwner() {
+        checkCallerOrThrow(Context.getOwner(), "OnlyOwner");
+    }
+
+    private void onlyAdmin() {
+        checkCallerOrThrow(this.admin(), "Only admin is allowed to call method");
+    }
+
+    private void onlyIBCHandler() {
+        checkCallerOrThrow(ibcHandler.get(), "Only IBCHandler allowed");
+    }
+
+//    private void checkService(String _svc) {
+//        Context.require(NAME.equals(_svc), "InvalidServiceName");
+//    }
+
+    private BigInteger getNextSn() {
+        BigInteger _sn = this.sn.getOrDefault(BigInteger.ZERO);
+        _sn = _sn.add(BigInteger.ONE);
+        this.sn.set(_sn);
+        return _sn;
+    }
+
+    private BigInteger getNextReqId() {
+        BigInteger _reqId = this.reqId.getOrDefault(BigInteger.ZERO);
+        _reqId = _reqId.add(BigInteger.ONE);
+        this.reqId.set(_reqId);
+        return _reqId;
+    }
+
+    private void cleanupCallRequest(BigInteger sn) {
+        requests.set(sn, null);
+    }
+
+
+    @Payable
+    @External
+    public BigInteger sendCallMessage(String _to, byte[] _data, @Optional byte[] _rollback) {
+        BigInteger sn = getNextSn();
+        Address caller = Context.getCaller();
+        Context.require(caller.isContract() || _rollback == null, "RollbackNotPossible");
+        Context.require(_data.length <= MAX_DATA_SIZE, "MaxDataSizeExceeded");
+        Context.require(_rollback == null || _rollback.length <= MAX_ROLLBACK_SIZE, "MaxRollbackSizeExceeded");
+
+        boolean needResponse = _rollback != null && _rollback.length > 0;
+
+        if (needResponse) {
+            CallRequest req = new CallRequest(caller, _to, _rollback);
+            requests.set(sn, req);
+        }
+
+        CSMessageRequest msgReq = new CSMessageRequest(caller.toString(), _to, sn, needResponse, _data);
+
+        Packet pct = new Packet();
+        pct.setSequence(sn);
```

Comment:
Sequence here is IBC specific, and while it should be fine for now like this. We might instead want to fetch this directly from the IBC contract:
   @External(readonly = true)
    public BigInteger getNextSequenceSend()

Reply:
Use the sequenceID from IBC (Host) itself?

Reply:
Yeah i don't see a reason not to really

Reply:
Used sequence ID from host in new commit